### PR TITLE
Make the name of the client certificate attribute which is mapped to roles configurable

### LINKED ISSUE #39364: Make the name of the client certificate attribute which is mapped to roles configurable
### Description

Currently, after #37269, it is possible to map the value of the certificate's `CN` attribute to the local roles.
@cescoffier has proposed to support other attributes such as the Subject Alternative Name (`SAN`) as `CN` may not  always be set.

### Implementation ideas

Add a property to let users customize the attribute name.
In meantime, it can be done with the custom `SecurityIdentityAugmentor`

### PR DESCRIPTION
**Pull request by michalvavrik**
closes: #39364

### CODE REVIEW

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/AuthRuntimeConfig.java
Comment (sberyozkin) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/AuthRuntimeConfig.java:49]: IMHO it should be a String, in case someone will want to refer to a custom attribute

Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/AuthRuntimeConfig.java:49]: > IMHO it should be a String, in case someone will want to refer to a custom attribute

we cannot really support reliably string for SAN, we can do it for custom DNs if you would like. I am not sure what the name should be as we will have 2 config properties and attribute is not correct name

Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/AuthRuntimeConfig.java:49]: @sberyozkin there is more than one way to specify SAN general name and I am not sure I want to write parser

File: integration-tests/mtls-certificates/src/main/resources/c-role-mappings.txt
Comment (sberyozkin) [inline integration-tests/mtls-certificates/src/main/resources/c-role-mappings.txt:1]: IMHO, it is quite unrealistic, which indirectly implies that capturing so many various attributes in the enum may be unnecessary

Comment (michalvavrik) [reply integration-tests/mtls-certificates/src/main/resources/c-role-mappings.txt:1]: I am fine with a String as long as for SAN it is a string for a known values (AKA enum :-)). Please suggest config property names because apart of fixing ASN1 parsing, I am ready to submit this PR for review. I am open to suggestion, just need something more concrete.

**Comment by michalvavrik** [comment]
There is difference betweenJDK 17 and JDK21  as far as SAN list is concerned (21 doesn't parse ASN1). I'll fix the CI failure.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/AuthRuntimeConfig.java
Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/AuthRuntimeConfig.java:49]: @michalvavrik I see, can you at least leave only those ones which might be used as mappers, and we can keep adding more based on some real cases.

Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/AuthRuntimeConfig.java:49]: truth be told @sberyozkin I originally had string but I didn't figure how to name this as I'd need 2 config properties and I'd prefer SAN to be enum and DN to be string. Maybe we could make it string with some values expected (SAN_xy) and some parsed DN_whatever? I really don't know.

File: integration-tests/mtls-certificates/src/main/resources/c-role-mappings.txt
Comment (michalvavrik) [reply integration-tests/mtls-certificates/src/main/resources/c-role-mappings.txt:1]: btw I agree that RNDs like `STREET` is quite unrealistic :-)

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/AuthRuntimeConfig.java
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/AuthRuntimeConfig.java:49]: thanks for feedback, let me try something in between...

File: integration-tests/mtls-certificates/src/main/resources/c-role-mappings.txt
Comment (michalvavrik) [reply integration-tests/mtls-certificates/src/main/resources/c-role-mappings.txt]: I dropped this test as you suggested it's unlikely scenario. I'll close this comment thread then.

**Comment by sberyozkin** [comment]
Hi @michalvavrik, it is done very comprehensively which is a trademark of your work, it is totally ready to go, and I think having String instead of the enum keeps options open for custom certificate attributes being used as well, though I see now it required you to do some extra parsing.

However, I'd like to suggest that you review which of SAN properties do we really want to support out of the box immediately. We don't usually do some wildcard kind of mappings from the identity to roles, as some of these attributes allow, like mapping any certificate from a given IP address to roles, etc. IMHO only those SAN attributes which can `uniquely` identify the subject should stay for now. At the very least, trimming  them a bit will likely simplify your code and we can grow it further. Think about it please

**Comment by michalvavrik** [comment]
> However, I'd like to suggest that you review which of SAN properties do we really want to support out of the box immediately. We don't usually do some wildcard kind of mappings from the identity to roles, as some of these attributes allow, like mapping any certificate from a given IP address to roles, etc. IMHO only those SAN attributes which can `uniquely` identify the subject should stay for now. At the very least, trimming them a bit will likely simplify your code and we can grow it further. Think about it please

Sure. I can see users mapping roles from:

- email (RFC822)
- URI (judging by Microsoft tags I googled https://techcommunity.microsoft.com/t5/ask-the-directory-services-team/preview-of-san-uri-for-certificate-strong-mapping-for-kb5014754/ba-p/3789785 they can be used for auth)
- other name (any)

@cescoffier @sberyozkin please let's share your thoughts on what SANs should be supported, it will limit number of iterations and reworks.

**Comment by sberyozkin** [comment]
I'd keep those 3 you listed, those which are used in some setups and are unique enough to identity a single subject

**Comment by michalvavrik** [comment]
> I'd keep those 3 you listed, those which are used in some setups and are unique enough to identity a single subject

I have dropped SANs but URI, ANY and RFC822. IMO we need to wait for @cescoffier feedback when he finds a time.

**Comment by sberyozkin** [comment]
Sounds good,  thanks @michalvavrik. Starting with these SAN attributes is reasonable IMHO

**Review by sberyozkin — APPROVED** [review]
LGTM, waiting for @cescoffier to review as well

**Comment by sberyozkin** [comment]
@michalvavrik By the way, I forgot to ask, can we use `CN` instead of `DN_CN`, the latter option reads quite technical, and clarity in docs it is part of the distinguished name ? So any attribute without `SAN` prefix or other extension 's prefix belongs to DN... It is not a big deal though,  I'll let you decide,  thanks

**Comment by michalvavrik** [comment]
hey @cescoffier , if you have any RFCs in the future when you find a time, just write them down and I'll address them ASAP. Thanks

**Pull request merged**
